swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift

```diff
@@ -6,90 +6,240 @@
+      kSecReturnPersistentRef: true,
+      kSecAttrAccessible: kSecAttrAccessibleAfterFirstUnlock,
+      // Specific to iOS:
+      kSecAttrAccessGroup: AppInfoPlistConstants.appGroupId as CFString as Any
+    ] as [CFString: Any]
+    #elseif os(macOS)
+    let query = [
+      // Common for both iOS and macOS:
+      kSecClass: kSecClassGenericPassword,
+      kSecAttrLabel: "Firezone access token (\(tokenAttributes.actorName))",
+      kSecAttrDescription: "Firezone access token",
+      kSecAttrService: tokenAttributes.authURLString,
+      kSecAttrAccount: "\(tokenAttributes.actorName): \(UUID().uuidString)", // The UUID uniquifies this item in the keychain
+      kSecValueData: token.data(using: .utf8) as Any,
+      kSecReturnPersistentRef: true,
+      kSecAttrAccessible: kSecAttrAccessibleAfterFirstUnlock,
+      // Specific to macOS:
+      kSecAttrAccess: try secAccessForAppAndNetworkExtension()
+    ] as [CFString: Any]
+    #endif
+    return try await withCheckedThrowingContinuation { [weak self] continuation in
+      self?.workQueue.async {
+        var ref: CFTypeRef?
+        let ret = SecStatus(SecItemAdd(query as CFDictionary, &ref))
+        guard ret.isSuccess else {
+          continuation.resume(throwing: KeychainError.appleSecError(call: "SecItemAdd", status: ret))
+          return
+        }
+        guard let savedPersistentRef = ref as? Data else {
+          continuation.resume(throwing: KeychainError.nilResultFromAppleSecCall(call: "SecItemAdd"))
+          return
+        }
+        // Remove any other keychain items for the same service URL
+        var checkForStaleItemsResult: CFTypeRef?
+        let checkForStaleItemsQuery = [
+          kSecClass: kSecClassGenericPassword,
+          kSecAttrService: tokenAttributes.authURLString,
+          kSecMatchLimit: kSecMatchLimitAll,
+          kSecReturnPersistentRef: true
+        ] as [CFString: Any]
+        let checkRet = SecStatus(SecItemCopyMatching(checkForStaleItemsQuery as CFDictionary, &checkForStaleItemsResult))
+        var isSavedItemFound = false
+        if checkRet.isSuccess, let allRefs = checkForStaleItemsResult as? [Data] {
+          for ref in allRefs {
+            if ref == savedPersistentRef {
+              isSavedItemFound = true
+            } else {
+              SecItemDelete([kSecValuePersistentRef: ref] as CFDictionary)
+            }
+          }
+        }
+        guard isSavedItemFound else {
+          continuation.resume(throwing: KeychainError.unableToFindSavedItem)
+          return
+        }
+        continuation.resume(returning: savedPersistentRef)
+      }
+    }
+  }
 
-    if status == Status.success {
-      return data as? Data
-    } else if status == Status.itemNotFound {
-      return nil
+  #if os(macOS)
+  private func secAccessForAppAndNetworkExtension() throws -> SecAccess {
+    // Creating a trusted-application-based SecAccess APIs are deprecated in favour of
```

Comment:
In the portal codebase we use the `XXX` convention for comments like these, where we've implemented something a certain way that needs to change or be refactored later.

Do you think it makes sense to use them for Apple too?

I'll need to update the contributing guide to note our convention:

- `FIXME`: broken functionality that should be fixed "soon" (within a few PRs or releases)
- `TODO`: Lower-priority refactoring or fixing
- `XXX`: Tech debt or something that's not ideal but should be revisited eventually

Reply:
I won't consider this tech debt because there's no alternate API for us to use, and we don't know yet of an alternate API that we can use in the future.

Yes, I'll try to adopt this convention.